universal-application-tool-0.0.1/app/services/applicant/Block.java

```diff
@@ -177,37 +177,52 @@ public boolean hasRequiredQuestionsThatAreUnansweredInCurrentProgram() {
   }
 
   /**
-   * Checks whether the block is complete - that is, {@link ApplicantData} has values at all the
-   * paths for all required questions in this block and there are no errors. Note: this cannot be
-   * memoized, since we need to reflect internal changes to ApplicantData.
+   * Checks whether the block is answered - that is, {@link ApplicantData} has values at all the
+   * paths for all questions in this block and there are no errors. Note: this cannot be memoized,
+   * since we need to reflect internal changes to ApplicantData.
+   *
+   * <p>A block with optional questions that are not answered is not answered.
```

Comment:
Thanks for this clarification. I'm starting to find it pretty confusing to track the meanings of "completed" and "answered" (and how they're used by higher level logic) across: 
- static questions
- optional questions
- questions answered in a previous program
- questions with errors
- and blocks with different combinations of these questions (e.g., optional-only block, static-only block)

It'd be helpful if we could find a place for a comment that gives a broad overview of the design.

Reply:
* Answered corresponds to whether the question has an answer, and completed is whether it is (answered) OR (optional and visited in this program but not answered).  Static questions evaluate as answered since they are not waiting for a response.
* For required questions, answered and completed will always be the same value.  For optional questions, the question can be completed but not answered if it was seen and passed over in that specific program for the applicant.
* Once a question is answered, it is always answered, regardless of which program it was answered in.
* Whether the question has errors is evaluated separately from whether it is answered or not.  That is an additional validation to know whether the block/application is ready to be submitted.
* Blocks evaluate on whether every question is answered/completed.  I.e. the block is only completed if ALL questions in it are also completed.  It doesn't matter what type of questions are in the block, only whether all of the questions have isCompleted = True.

I believe a write up of this is captured in the ApplicantQuestion javadocs, and then a block-specific expansion here in the Block javadocs.  Let me know if there is anything else specifically you see is missing.